Extraction of macro definitions from callstacks

PLukas2018 opened the pull request:
Second part of commits from https://github.com/diffkemp/diffkemp/pull/314 .
This PR adds:
- method for extracting macro definitions from the call stacks (it is possible to extract information about definitions for all macros except the differing macros for which is the definition extracted in [#333](https://github.com/diffkemp/diffkemp/pull/333) PR)
- update of `get_end_line` function to be possible to use it also for finding line where macro ends.

viktormalik commented on diffkemp/semdiff/result.py:
```diff
@@ -81,6 +81,17 @@ def to_output_yaml_with_rel_path(self, prefix):
+        @staticmethod
+        def get_name_and_kind(call):
+            """Returns tuple (name, kind) for a call."""
+            name = call["name"]
+            kind = "function"
+            if " " in name:
+                name, kind = name.split(" ")
+                # kind looks like this "(kind)", extracting it from brackets
+                kind = kind[1:-1]
+            return (name, kind)
```
What about this? Slightly shorter and more readable (at least for me)
```suggestion
            split = call["name"].split()
            if len(split) == 1:
                return (split[0], "function")
            # The expected form is "name (kind)"
            return (split[0], split[1][1:-1])
```
Also note that whitespace is the default separator for `split`.

viktormalik commented on diffkemp/utils.py:
```diff
@@ -53,25 +53,29 @@ class EndLineNotFound(Exception):
+        while not is_end(line):
             line_number += 1
-            if line_number > len(lines):
+            if line_number > lines_amount:
```
Why not use `len(lines)` directly? It's a constant function so there's no performance lost.

PLukas2018 replied:
Thanks, I did not know that.

viktormalik commented on diffkemp/semdiff/result.py:
```diff
@@ -120,6 +120,46 @@ def get_symbol_names(self, compared_fun):
+        def get_macro_defs(self, compared_fun):
+            """
+            Returns a tuple:
+            - list of macro definitions (except last macro in the callstack)
```
I'm wondering if we should call this "macro locations" rather than "macro definitions". The latter implies to me that we're extracting macro bodies while in fact, we're just extracting locations where the macros start. OTOH, if we're using the word "definition" in this context for functions, then it's probably better to stay consistent.

PLukas2018 replied:
Good point, we are already using the word 'definition' eg. here https://github.com/diffkemp/diffkemp/blob/4b94c9acd38909fc5b934dabde188bd15135e05b/diffkemp/output.py#L76-L79 , so I would probably stick with the word 'definition'.

viktormalik commented on diffkemp/semdiff/result.py:
```diff
@@ -120,6 +120,46 @@ def get_symbol_names(self, compared_fun):
+            # For getting last macro deffinition
+            last_macro_name = None
+            last_function_name = compared_fun
+            if self.calls:
```
I'd invert this to save some indentation.

viktormalik reviewed (commented):
LGTM, just a couple of code styling remarks (feel free to ignore those that you don't like).

Also one more thing: could you try to add a short description to each commit? Even if the commit is simple (like a refactoring), it's always good to explain what's done there as it gives the author's intent and therefore makes the commit easier to understand.

For instance, for your commits no.1 and 3, there could be something like:
```
Result Callstack refactoring

Extract splitting symbol name from a symbol kind to a separate function.
Also invert a condition to decrease indentation.
```
```
Add getting end line for macros

The end line of a macro definition is the line which does not end with
an escape backslash ('\').
```

PLukas2018 commented:
Thanks for the review, I resolved the comments.

Also, I fixed a mistake that I made in `get_end_line` function - because `readlines()` returns a newline character it is necessary always to use `rstrip()` (even for checking end line in macros). 
I also added unit tests for the `get_end_line` function.

viktormalik reviewed (approved):
Nice!
